Document scale transformations in the PredEvals user guide (#477)
hubverse-org/hubDocs: docs/source/user-guide/dashboards.md

:::

(predevals-scale-transformations)=
### Scale transformations (optional)

When a target's values span a wide range of magnitudes, scores computed on the natural scale can be dominated by the largest values. Applying a scale transformation before scoring can give a more balanced evaluation across magnitudes. PredEvals supports configuring scale transformations in `predevals-config.yml` from schema version `v1.1.0` onwards, applied instead of or alongside natural-scale scoring.

Scale transformations require schema version `v1.1.0` or later. If your config's `schema_version` points to an earlier version, update it to a version that supports transforms — for example, the one that introduced them:

```yaml
schema_version: https://raw.githubusercontent.com/hubverse-org/hubPredEvalsData/main/inst/schema/v1.1.0/config_schema.json
```

Transforms can be configured at two levels:

- **`transform_defaults`** (top-level, optional): a default transformation applied to every target whose available output types support transformation.

> - **`targets[*].transform`** (per-target, optional): a transformation specific to one target. A per-target `transform` *replaces* `transform_defaults` entirely for that target.

Reviewer: I don't really understand the syntax with the asterisk inside the brackets for the per-target transform field, and it wasn't clear to me what this looked like until an example config was shown further done on the page. Can we put an abbreviated toy config right below this block for clarity?

Author: Good call — the `targets[*].transform` notation was JSON-path jargon. I've replaced it with plain wording ("a `transform` field on an individual target") and added an abbreviated toy config right below the two-level explanation, showing all three behaviours: a target inheriting `transform_defaults`, one overriding it with its own transform, and one opting out via `transform: false`. Pushed in 38bf272.

rounds_idx: 0
transform_defaults:
  fun: log_shift
  args:
    offset: 1
targets:
- target_id: wk inc flu hosp
  metrics:
  - wis
  - ae_median
- target_id: wk flu hosp rate category
  metrics:
  - log_score
  - rps
  transform: false
```

In this example, `wk inc flu hosp` inherits `log_shift` (with `offset: 1`) from `transform_defaults`. The categorical target `wk flu hosp rate category` opts out explicitly with `transform: false`.

(predevals-transform-output)=
#### How transformed scores appear in the dashboard

When a transform applies to a target, the dashboard presents the transformed-scale metrics labelled with the transform's `label` in parentheses — for example, `WIS (log)` alongside `WIS`, or `Rel. WIS (log)` alongside `Rel. WIS`.

- With **`append: true`** (the default), both the natural-scale and transformed-scale version of each affected metric are available.
- With **`append: false`**, only the transformed-scale version is shown.

Two metric families — interval coverage (for example `50% Cov.`, `95% Cov.`) and bias — are mathematically unchanged by any monotonic transform, so they are shown only once, with no separate transformed-scale variant.

> 

Reviewer: I think it would be beneficial to include a screenshot of what this actually looks like on the dashboard evaluations page, like the screenshots provided earlier on the page.

Author: Agreed that a screenshot would help. The catch is the predevals dashboard UI for transforms isn't finalised yet (this PR is intentionally landing ahead of the v1.1.0 release), so there's no real view to capture — I'd rather not mock one up and risk it not matching the shipped UI. I've added a `TODO` placeholder comment at that spot so we add a real screenshot (matching the other eval-page screenshots on this page) once the UI lands. Pushed in 38bf272 — does that work as a follow-up?

Reviewer: Yes, that's totally fine, thank you!

:::

(predevals-scale-transformations)=

> ### Scale transformations (optional)

Reviewer 2: I would suggest adding a word that clarifies this is for evaluations. When I started reading this, I thought it referred to transformation of scores, like counts. Although it may be obvious to people who read this whole page from start to finish, it is less so, for people who are quickly skimming the manual (as most of us do). Maybe the title below, or something like it?

```suggestion
### Scale transformations for PredEvals (optional)
```

Author: Agreed — scoping it to PredEvals in the heading helps skimmers land in the right place. Updated the heading to **"Scale transformations for PredEvals (optional)"**.

> 

Reviewer 2: This is not specific to the new content, but in general, in this section, and in fact in none of the hubDocs, was I able to find an explanation of some of the metrics. Specifically, I was searching for what the "ae" stood for in `ae_median`. I searched in hubDocs, and not finding anything, went to Google. Reading "absolute error" made it seem obvious, but it would be nice for it to be explicit in this documentation. Or at the very least, to have a pointer to where we can read those definitions (I didn't find one). In yesterday's presentation I saw how the dashboards now have "metric definitions", could we include something like that, here? Or at least a link to learn more?

Author: Good point that the abbreviations deserve a definition. The page already introduces these a bit earlier — the worked example at the top of the PredEvals section spells out "the weighted interval score (WIS), absolute error of the median…" — and I've expanded the link below the output-types table to point explicitly at the [hubEvals `score_model_out()` reference](https://hubverse-org.github.io/hubEvals/reference/score_model_out.html#details) (e.g. what `wis` / `ae_median` mean) and the [scoringutils scoring-rules article](https://epiforecasts.io/scoringutils/articles/scoring-rules.html). We're deliberately keeping the metric definitions in hubEvals/scoringutils rather than copying them here — these are the dashboard-specific docs, and the metric details track scoringutils, so a second copy would just be another place to keep in sync.

When a target's values span a wide range of magnitudes, scores computed on the natural scale can be dominated by the largest values. Applying a scale transformation before scoring can give a more balanced evaluation across magnitudes. PredEvals supports configuring scale transformations in `predevals-config.yml` from schema version `v1.1.0` onwards, applied instead of or alongside natural-scale scoring.

Scale transformations require schema version `v1.1.0` or later. If your config's `schema_version` points to an earlier version, update it to a version that supports transforms — for example, the one that introduced them:

```yaml
schema_version: https://raw.githubusercontent.com/hubverse-org/hubPredEvalsData/main/inst/schema/v1.1.0/config_schema.json
```

Transforms can be configured at two levels:

- **`transform_defaults`** (top-level, optional): a default transformation applied to every target whose available output types support transformation.
- A **`transform`** field on an individual target (per-target, optional): a transformation specific to that one target, which *replaces* `transform_defaults` entirely for it.

> For example, the abbreviated config below sets a default transform, has the first target inherit it, the second target use its own transform instead, and the third opt out:

Reviewer 2: As someone who doesn't use this, I would edit this slightly to clarify that you will provide details on what goes in this example, below.

```suggestion
Details regarding the structure and functions available for transforms will be provided below, but it may be easier to start with an illustrative case. In the example below, the abbreviated config sets a default transform, has the first target inherit it, the second target use its own transform instead, and the third opt out:
```

Author: Good suggestion — I've reworked the lead-in so it signals the structure and available functions are detailed below, and frames the config as a starting illustration. Thanks!

```yaml
transform_defaults:          # applies to all transformable targets by default
  fun: log_shift
  args:

>     offset: 1

Reviewer 2: Sorry if I'm being pedantic, but could you choose a different offset than "1" (maybe 0.5, or 10)? The reason is that you can use `log1p` instead, so why would you do it the long way of writing down the offset.

Author: I'd actually like to keep `offset: 1` here. The purpose of this example is to show how `args` are passed to a `fun`, and `log(x + 1)` is one of the most common, realistic choices for count data that can hit zero. The fact that `log1p` happens to be an equivalent shortcut doesn't really change what the example is teaching — `log_shift` is the general, flexible function and `args` is the thing being demonstrated. So I've left it as `offset: 1`. Appreciate the close read though!

targets:
- target_id: wk inc flu hosp
  metrics: [wis]
  # no transform field -> inherits transform_defaults (log_shift)
- target_id: wk inc covid hosp

>   metrics: [wis]

Reviewer 2: Just a thought, but as I was thinking about this example, I wondered why you would apply different transformations for the same metric, but with different `target_id`s (it sounded fishy). Maybe this is from a use case, and so my comment is wrong, but for this second case, could you perhaps do `wk inc flu death` and `[ae_median]`? Or maybe that example is also wrong?

Author: Thanks for flagging this — it's a fair instinct. The repeated `wis` is actually deliberate: I wanted the *only* thing varying between the first two targets to be the `transform` field, so the example reads purely as "this is how a per-target transform overrides the default," independent of which metric is in play. If I changed both the target and the metric, it could imply the transform choice is somehow tied to the metric, which isn't the case. So I've kept it as-is, but let me know if you feel it still reads as misleading.

| `log1p`      | [base R](https://rdrr.io/r/base/Log.html) | Natural logarithm of `(1 + x)`, i.e. `log(1 + x)`. | — |
| `log`        | [base R](https://rdrr.io/r/base/Log.html) | Natural logarithm (base *e*). Undefined at `0`. | `base` (default *e*). |
| `log10`      | [base R](https://rdrr.io/r/base/Log.html) | Base-10 logarithm. | — |
| `log2`       | [base R](https://rdrr.io/r/base/Log.html) | Base-2 logarithm. | — |
| `sqrt`       | [base R](https://rdrr.io/r/base/MathFun.html) | Square root. | — |

#### Which output types are transformed

A target's single resolved transform applies uniformly to **every transformable output type** the target declares. The transformable output types are:

| Output type | Transformable? |
| ----------- | -------------- |
| `mean`, `median`, `quantile` | Yes |
| `sample` | Not yet — support is under consideration |
| `pmf` | No (categorical output types are not meaningfully transformed) |

> Metrics computed on a non-transformable output type (for example, the pmf metrics `log_score` or `rps`) are always scored on the natural scale and receive no transformed-scale column. See the [hubEvals metric × output-type compatibility reference](https://hubverse-org.github.io/hubEvals/reference/score_model_out.html#details) for which metrics apply to which output types.

Reviewer 2: ```suggestion
Metrics computed on a non-transformable output type (for example, the pmf metrics `log_score` or `rps`) are always scored on the natural scale and receive no transformed-scale column. See the [hubEvals `score_model_out` documentation details](https://hubverse-org.github.io/hubEvals/reference/score_model_out.html#details) to see what metrics (e.g., `wis`) can be applied to each `output-type` (e.g., `quantile`).
```

Author: Adopted, thank you — that's much more concrete. I've reworded it to point at the `score_model_out()` details for the definition of each metric (e.g. what `wis` or `ae_median` mean) and which metrics can be applied to each output type. This also doubles as the pointer for your metric-definitions comment above.
